File: BM4_合并两个排序的链表/java_solutions/IterationSolution.java
package BM4_合并两个排序的链表.java_solutions;

public class IterationSolution {
    /**
     * 输入两个递增的链表，单个链表的长度为n，合并这两个链表并使新链表中的节点仍然是递增排序的。
     *
     * 
     * @param pHead1 ListNode类 
     * @param pHead2 ListNode类 
     * @return ListNode类
     */
    public ListNode Merge (ListNode pHead1, ListNode pHead2) {
        ListNode resultHead = new ListNode(-1);
        ListNode current = resultHead;
        while(pHead1 != null && pHead2 != null) {
            if(pHead1.val <= pHead2.val){
                current.next = pHead1;
                pHead1 = pHead1.next;
            }else{
                current.next = pHead2;
                pHead2 = pHead2.next;
            }

            current = current.next;
        }

        current.next = pHead1 != null ? pHead1 : pHead2;

        return resultHead.next;
    }
}

File: BM4_合并两个排序的链表/javascript_solutions/iteration_solution.ts
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

/**
 * 输入两个递增的链表，单个链表的长度为n，合并这两个链表并使新链表中的节点仍然是递增排序的。
 *
 * 
 * @param pHead1 ListNode类 
 * @param pHead2 ListNode类 
 * @return ListNode类
 */
export function Merge(pHead1: ListNode | null, pHead2: ListNode | null): ListNode | null {
    const resultHead = new ListNode(-1);

    let current = resultHead;
    let head1: ListNode | null = pHead1;
    let head2: ListNode | null = pHead2;

    while(head1 && head2){
        if(head1.val <= head2.val){
            current.next = head1;
            head1 = head1.next;
        }else{
            current.next = head2;
            head2 = head2.next;
        }
        
        current = current.next;
    }
    current.next = head1 ? head1 : head2;

    return resultHead.next;
}
